### PR TITLE
Avoid panics

### PR DESCRIPTION
**Pull request by dhontecillas**
Reduce the chance of panicking when a sensible behavior can be used, while emitting an error.

### CODE REVIEW

File: proxy/concurrent.go [hunk 1]
```diff
@@ -8,18 +8,19 @@ import (
...
+	"github.com/luraproject/lura/v2/logging"
 )
 
-// NewConcurrentMiddleware creates a proxy middleware that enables sending several requests concurrently
-func NewConcurrentMiddleware(remote *config.Backend) Middleware {
+// NewConcurrentMiddlewareWithLogger creates a proxy middleware that enables sending several requests concurrently
+func NewConcurrentMiddlewareWithLogger(logger logging.Logger, remote *config.Backend) Middleware {
 	if remote.ConcurrentCalls == 1 {
-		panic(ErrTooManyProxies)
+		logger.Fatal("too few concurrent calls: NewConcurrentMiddleware expects more than 1 concurrent call, got %d", remote.ConcurrentCalls)
```
Comment (kpacha) [inline proxy/concurrent.go:17]: add a return for better testing

File: proxy/concurrent.go [hunk 2]
```diff
@@ -8,18 +8,19 @@ import (
...
 	}
 	serviceTimeout := time.Duration(75*remote.Timeout.Nanoseconds()/100) * time.Nanosecond
 
 	return func(next ...Proxy) Proxy {
 		if len(next) > 1 {
-			panic(ErrTooManyProxies)
+			logger.Fatal("too many proxies for this proxy middleware: NewConcurrentMiddleware only accepts 1 proxy, got %d", len(next))
```
Comment (kpacha) [inline proxy/concurrent.go:23]: add a return for better testing

File: proxy/formatter.go [hunk 3]
```diff
@@ -320,8 +320,8 @@ func newFlatmapFormatter(cfg config.ExtraConfig, target, group string) *flatmapF
 func NewFlatmapMiddleware(logger logging.Logger, cfg *config.EndpointConfig) Middleware {
 	formatter := newFlatmapFormatter(cfg.ExtraConfig, "", "")
 	return func(next ...Proxy) Proxy {
-		if len(next) != 1 {
-			panic(ErrTooManyProxies)
+		if len(next) > 1 {
+			logger.Fatal("too many proxies for this proxy middleware: NewFlatmapMiddleware only accepts 1 proxy, got %d", len(next))
```
Comment (kpacha) [inline proxy/formatter.go:324]: add a return for better testing

File: proxy/graphql.go [hunk 4]
```diff
@@ -63,12 +63,12 @@ func NewGraphQLMiddleware(logger logging.Logger, remote *config.Backend) Middlew
...
 
 	default:
-		return EmptyMiddleware
+		return emptyMiddlewareFallback(logger)
 	}
 
 	return func(next ...Proxy) Proxy {
 		if len(next) > 1 {
-			panic(ErrTooManyProxies)
+			logger.Fatal("too many proxies for this proxy middleware: NewGraphQLMiddleware only accepts 1 proxy, got %d", len(next))
```
Comment (kpacha) [inline proxy/graphql.go:71]: add a return for better testing

File: proxy/headers_filter.go [hunk 5]
```diff
@@ -11,14 +11,14 @@ import (
...
+func NewFilterHeadersMiddleware(logger logging.Logger, remote *config.Backend) Middleware {
 	if len(remote.HeadersToPass) == 0 {
-		return EmptyMiddleware
+		return emptyMiddlewareFallback(logger)
 	}
 
 	return func(next ...Proxy) Proxy {
 		if len(next) > 1 {
-			panic(ErrTooManyProxies)
+			logger.Fatal("too many proxies for this proxy middleware: NewFilterHeadersMiddleware only accepts 1 proxy, got %d", len(next))
```
Comment (kpacha) [inline proxy/headers_filter.go:21]: add a return for better testing

File: proxy/logging.go [hunk 6]
```diff
@@ -15,7 +15,7 @@ func NewLoggingMiddleware(logger logging.Logger, name string) Middleware {
 	logPrefix := "[" + strings.ToUpper(name) + "]"
 	return func(next ...Proxy) Proxy {
 		if len(next) > 1 {
-			panic(ErrTooManyProxies)
+			logger.Fatal("too many proxies for this proxy middleware: NewLoggingMiddleware only accepts 1 proxy, got %d", len(next))
```
Comment (kpacha) [inline proxy/logging.go:18]: add a return for better testing

File: proxy/merging.go [hunk 7]
```diff
@@ -39,6 +42,9 @@ func NewMergeDataMiddleware(logger logging.Logger, endpointConfig *config.Endpoi
 
 	return func(next ...Proxy) Proxy {
 		if len(next) != totalBackends {
+			// we leave the panic here, because we do not want to continue
+			// if this configuration is wrong, as it would lead to unexpected
+			// behaviour.
 			panic(ErrNotEnoughProxies)
```
Comment (kpacha) [inline proxy/merging.go:48]: replace with a log.Fatal and a return

File: proxy/plugin.go [hunk 8]
```diff
@@ -88,7 +88,8 @@ func newPluginMiddleware(logger logging.Logger, tag, pattern string, cfg map[str
 
 	return func(next ...Proxy) Proxy {
 		if len(next) > 1 {
-			panic(ErrTooManyProxies)
+			logger.Fatal("too many proxies for this proxy middleware: newPluginMiddleware only accepts 1 proxy, got %d tag: %s, pattern: %s",
+				len(next), tag, pattern)
```
Comment (kpacha) [inline proxy/plugin.go:92]: add a return for better testing

**Review by kpacha — CHANGES_REQUESTED** [review]

**Comment by dhontecillas** [comment]
Fixed code review issues.

**Review by kpacha — APPROVED** [review]

**Pull request merged**